pkg/trace/buffered_file.go

```diff
@@ -2,55 +2,102 @@ package trace
 
 import (
 	"bufio"
+	"errors"
+	"io"
 	"os"
 	"sync"
+	"sync/atomic"
 )
 
+// bufferedFile is a file that is being written to and read from. It is thread
+// safe, however, when reading from the file, writes will be ignored.
 type bufferedFile struct {
-	mut *sync.RWMutex
+	// reading protects the file from being written to while it is being read
+	// from. This is needed beyond in addition to the mutex so that writes can
+	// be ignored while reading.
+	reading atomic.Bool
+
+	// mut protects the buffered writer.
+	mut *sync.Mutex
+
 	// file is the file that is being written to.
 	file *os.File
+
 	// writer is the buffered writer that is writing to the file.
 	wr *bufio.Writer
 }
 
 // newbufferedFile creates a new buffered file that writes to the given file.
 func newbufferedFile(file *os.File) *bufferedFile {
 	return &bufferedFile{
-		mut:  &sync.RWMutex{},
-		file: file,
-		wr:   bufio.NewWriter(file),
+		file:    file,
+		wr:      bufio.NewWriter(file),
+		reading: atomic.Bool{},
+		mut:     &sync.Mutex{},
 	}
 }
 
-// Write writes the given bytes to the file.
+// Write writes the given bytes to the file. If the file is currently being read
+// from, the write will be lost.
 func (f *bufferedFile) Write(b []byte) (int, error) {
+	if f.reading.Load() {
+		return 0, nil
+	}
```

Comment:
re-iterating an oddity for reviewers to hopefully help get this across the line. This seems silly to have an atomic and a mutex, however if we want to have the ability to ignore writes while also avoiding all races, then we need both the atomic and the mutex